plugin/go.vim

```diff
@@ -260,6 +260,16 @@ augroup vim-go
   " in the same window doesn't highlight the most recently matched
   " identifier's positions.
   autocmd BufWinEnter *.go call go#guru#ClearSameIds()
-augroup END
+
+  autocmd BufEnter *.go
+        \  if get(g:, 'go_autodetect_gopath', 0)
+        \|   let b:old_gopath = $GOPATH
+        \|   let $GOPATH = go#path#Detect()
+        \| endif
+  autocmd BufLeave *.go
+        \  if get(g:, 'go_autodetect_gopath', 0)
+        \|   let $GOPATH = b:old_gopath
```

Comment:
(nit) When the old `GOPATH` was not set, restoring `GOPATH` to an empty string does not faithfully restore the state to what it was. As a pragmatic matter, all the go tooling will handle this case correctly, but users may have other scripts, etc. that do not. As I see it, vim-go should restore `GOPATH` to its state prior to entering the buffer.